File: csharp/1493.longest-subarray-of-1-s-after-deleting-one-element.cs
/*
 * @lc app=leetcode id=1493 lang=csharp
 *
 * [1493] Longest Subarray of 1's After Deleting One Element
 */

using System;

// @lc code=start
public class Solution {
    public int LongestSubarray(int[] nums) {
        int pi = -1, pc = 0, ret = 0, n = nums.Length;
        for (int i = 0; i < n;) {
            if (nums[i] != 1) {
                ++i;
                continue;
            }
            int ni = i + 1;
            while (ni < n && nums[ni] == 1) {
                ++ni;
            }
            if (i == 0 && ni == n) {
                return n - 1;
            }
            ret = Math.Max(ret, ni - i);
            if (pi + pc == i - 1) {
                ret = Math.Max(ret, ni - i + pc);
            }
            pi = i;
            pc = ni - i;
            i = ni;
        }
        return ret;
    }
}
// @lc code=end

File: java/1493.longest-subarray-of-1-s-after-deleting-one-element.java
/*
 * @lc app=leetcode id=1493 lang=java
 *
 * [1493] Longest Subarray of 1's After Deleting One Element
 */

// @lc code=start
class Solution {
    public int longestSubarray(int[] nums) {
        int n = nums.length, acc = 0, max = 0, ret = 0, prevAcc = 0;
        for (int i = 0; i <= n; ++i) {
            if (i < n && nums[i] == 1) {
                ++acc;
                continue;
            }
            if (i > 0 && nums[i - 1] == 1) {
                ret = Math.max(acc + prevAcc, ret);
                max = Math.max(acc, max);
                prevAcc = acc;
                acc = 0;
            } else {
                prevAcc = 0;
            }
        }
        if (max == n) {
            return n - 1;
        }
        return Math.max(max, ret);
    }
}
// @lc code=end

